chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+	getEventsFnc := func(stmt *sql.Stmt, values []any) ([]*CollectedEvent, error) {
+		q, err := stmt.QueryContext(ctx, values...)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to query events: %w", err)
+		}
+		defer func() { _ = q.Close() }()
+
+		var ces []*CollectedEvent
+		var currentID int64 = -1
+		var ce *CollectedEvent
+
+		for q.Next() {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+
+			var row struct {
+				id           int64
+				height       uint64
+				tipsetKeyCid []byte
+				emitterID    uint64
+				emitterAddr  []byte
+				eventIndex   int
+				messageCid   []byte
+				messageIndex int
+				reverted     bool
+				flags        []byte
+				key          string
+				codec        uint64
+				value        []byte
+			}
+
+			if err := q.Scan(
+				&row.id,
+				&row.height,
+				&row.tipsetKeyCid,
+				&row.emitterID,
+				&row.emitterAddr,
+				&row.eventIndex,
+				&row.messageCid,
+				&row.messageIndex,
+				&row.reverted,
+				&row.flags,
+				&row.key,
+				&row.codec,
+				&row.value,
+			); err != nil {
+				return nil, xerrors.Errorf("read prefill row: %w", err)
+			}
+
+			if row.id != currentID {
+				if ce != nil {
+					ces = append(ces, ce)
+					ce = nil
+					// Unfortunately we can't easily incorporate the max results limit into the query due to the
+					// unpredictable number of rows caused by joins
+					// Break here to stop collecting rows
+					if f.MaxResults > 0 && len(ces) >= f.MaxResults {
+						break
+					}
+				}
+
+				currentID = row.id
+				ce = &CollectedEvent{
+					EventIdx: row.eventIndex,
+					Reverted: row.reverted,
+					Height:   abi.ChainEpoch(row.height),
+					MsgIdx:   row.messageIndex,
+				}
+
+				if row.emitterAddr == nil {
+					ce.EmitterAddr, err = address.NewIDAddress(row.emitterID)
+					if err != nil {
+						return nil, xerrors.Errorf("failed to parse emitter id: %w", err)
+					}
+				} else {
+					ce.EmitterAddr, err = address.NewFromBytes(row.emitterAddr)
+					if err != nil {
+						return nil, xerrors.Errorf("parse emitter addr: %w", err)
+					}
+				}
+
+				tsKeyCid, err := cid.Cast(row.tipsetKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse tipsetkey cid: %w", err)
+				}
+
+				ts, err := si.cs.GetTipSetByCid(ctx, tsKeyCid)
+				if err != nil {
+					return nil, xerrors.Errorf("get tipset by cid: %w", err)
+				}
+
+				ce.TipSetKey = ts.Key()
+
+				ce.MsgCid, err = cid.Cast(row.messageCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse message cid: %w", err)
+				}
+			}
+
+			ce.Entries = append(ce.Entries, types.EventEntry{
+				Flags: row.flags[0],
+				Key:   row.key,
+				Codec: row.codec,
+				Value: row.value,
+			})
+		}
+
+		if ce != nil {
+			ces = append(ces, ce)
+		}
+
+		if len(ces) == 0 {
+			return nil, nil
+		}
+
+		// collected event list is in inverted order since we selected only the most recent events
```

Comment:
why do we do this? is it to do with MaxResults?

Reply:
The sort is return events in ascending order of height.

Reply:
I'm mot interested in why our query isn't doing this work for us?

Reply:
@rvagg 

So the query does sort by both height and insertion order of event entries:
```
s := `SELECT
			e.event_id,
			tm.height,
			tm.tipset_key_cid,
			e.emitter_id,
			e.emitter_addr,
			e.event_index,
			tm.message_cid,
			tm.message_index,
			e.reverted,
			ee.flags,
			ee.key,
			ee.codec,
			ee.value
		FROM event e
		JOIN tipset_message tm ON e.message_id = tm.message_id
		JOIN event_entry ee ON e.event_id = ee.event_id`

	if len(joins) > 0 {
		s = s + ", " + strings.Join(joins, ", ")
	}

	if len(clauses) > 0 {
		s = s + " WHERE " + strings.Join(clauses, " AND ")
	}

	// retain insertion order of event_entry rows
	s += " ORDER BY tm.height DESC, ee._rowid_ ASC"
	return values, s, nil
```

But not sure why we then sort ascending just by height.

**_However,_** this code is from the event index and is working really well now for ETH tooling/ETH RPC and I don't want to change it unless there is a strong use case/user motivation to do so. Otherwise, we'll have to profile it all over again/ensure it is compatible with ETH tooling etc. 

Let's keep it.

Reply:
ok, fair call